covid_graphs/covid_graphs/formula.py

```diff
@@ -1,81 +1,90 @@
+        label: str,
+    ) -> None:
+        """
+        TODO
+        [start_date, end_date)
+        """
+        self._func = func
+        self.start_date = start_date
+        self.end_date = end_date
+        self.label = label
+        self._xs: Optional[List[datetime.date]] = None
+        self._ys: Optional[List[float]] = None
 
-def ATG_formula(TG, A, exponent, min_case_count=200):
-    text = r"$\frac{_A}{_TG} \cdot \left(\frac{t}{_TG}\right)^{_expon} / e^{t/_TG}$"
-    text = text.replace("_A", f"{A:.0f}").replace("_TG", f"{TG}").replace("_expon", f"{exponent}")
-    # Day of the second inflection point
-    second_ip_day = math.ceil(TG * (exponent + math.sqrt(exponent)))
-    return Formula(
-        lambda t: (A / TG) * (t / TG) ** exponent / np.exp(t / TG),
-        text,
-        second_ip_day,
-        min_case_count,
-    )
+    def get_trace(
+        self, plot_start_date: datetime.date, plot_end_date: datetime.date
+    ) -> Tuple[List[datetime.date], List[float]]:
+        """
+        Returns (xs, ys) - a trace of the function `func` in the given
+        half-open interval [plot_start_date, plot_end_date).
+        """
+        self._xs = [
+            plot_start_date + datetime.timedelta(days=d)
+            for d in range((plot_end_date - plot_start_date).days)
+        ]
+        self._ys = [self._func(x) for x in self._xs]
+        return self._xs, self._ys
 
+    def get_maximum(self) -> Tuple[datetime.date, float]:
+        """
+        Returns (x, y) point on the curve with maximal y.
+        """
+        if self._xs is None or self._ys is None:
+            raise ValueError("Must call get_trace() first.")
+        idx = np.argmax(self._ys)
+        return self._xs[idx], self._ys[idx]
 
-def first_day_of_curve(cumulative_active, formula):
-    """
-    Returns the first day of the curve formed by 'formula'.
 
-    It's the first day for which the number of cumulative active is more than
-    formula.min_case_count.
-    """
-    return np.argmax(cumulative_active > formula.min_case_count)
+class CurveConstructor:
+    @abstractmethod
+    def get_curve(self, country_report: CountryReport) -> Curve:
+        pass
 
 
-class Curve:
-    """
-    Class containing a curve derived from a formula.
+@dataclass
+class AtgCurveConstructor(CurveConstructor):
+    tg: float
+    a: float
+    exponent: float
+    min_case_count: int
 
-    cumulative_active - list of cumulative active cases
-    first_idx - 0-based index where the graph starts, so from cumulative_active[first_idx:].
-    last_idx - 0-based index where the graph ends
-    first_date - date corresponding to cumulative_active[0]
-    """
+    def get_curve(self, country_report: CountryReport) -> Curve:
+        label = r"$\frac{_A}{_TG} \cdot \left(\frac{t}{_TG}\right)^{_expon} / e^{t/_TG}$"
+        label = (
+            label.replace("_A", f"{self.a:.0f}")
+            .replace("_TG", f"{self.tg}")
+            .replace("_expon", f"{self.exponent}")
```

Comment:
This is actually a hack that I quickly made. f-strings and TeX formulas don't mix well. Is there a better way than `replace`?

Reply:
Not that I know of, maybe there is :)